### ISSUE TITLE
Subspace relays

### ISSUE THREAD

**Issue by nosknut** [LemADEC/WarpDrive#284]
A galactic public network of messages. Possibly a block that computers can interact with. Basically rednet but all relays see the same and it has a huge range ...

**Comment by nosknut** [comment]
Ps. This is a suggestion for a new feature

**Comment by LemADEC** [comment]
What kind of message are we talking about?
There would be a limited length and some transmission delay to account for distance.
We could require the area to be covered by a long range antenna of sort.

**Comment by StargateMC** [comment]
Haha I was going to do this as an OC add-on. Doesn't really need to be WD.
He refers to how Subspace transmissions work in Stargate. Power is a factor, range is a factor, delay is not... not that you couldn't add something in there.

**Comment by LemADEC** [comment]
At least 1 tick delay is required due to multithreading involved with CC/OC.
A good second delay is a good way to prevent players abusing it and killing server CPU through it.

**Comment by StargateMC** [comment]
Fair enough :)

**Comment by nosknut** [comment]
U wouldent need it to be covered by an antenna. Shure u could have different tiers to account for longrer distances, but the idea is that it can reach galactic distances like 80k block distance. The point of a subspace relay is the vast distance becus its transmitted tru hyperspace.

**Comment by LemADEC** [comment]
Area restriction doesn't really make sense. Better be a block you connect to that will transmit the message with some delay & power consumption.
Normal is same dimension, Advanced is same star system (same space dimension), Superior is unlimited.
Emitting and receiving blocks need to respect the tiers constrains (or just the emitter?).

Message is a timestamp, a header (16 or 32 bits unsigned integer) followed by a separator (':' or ' '?) and a content (250 chars).
Message origin remains unknown.

Same block is used for emission and reception (?).
Block can be setup to filter reception on message header (only one header).
Block shall be placed for x ticks before it's 'tuned' to the subspace network. 5 to 20 mn?
Block shall remain powered for x ticks before it's 'warmed up'. 5 to 20 s?

Emitting block consumes energy per message sent, function of tier (not of distance or number of receivers, it's just an UDP broadcast).
Actual emission happens with next tick. First message given is the one sent. A failure is returned if a message is already queued.

Receiving block consumes energy per messages queued, only counting the ones matching the filter. (i.e. you can't kill it by spamming the network)

Block remember its messages queues when losing power.
Block loose tuning, energy and message queue when broken. Filter is kept?

Receiving block shall be loaded and tuned when message is sent.
Receiving block can queue up to (1 + 1 per Memory crystal upgrades) messages.
In case of overflow, most recent message is replaced with "-overflow detected-" and message header is kept. Nota: there's no hint on how many messages were lost.
Every x ticks, receiving block post the oldest message from the queue as an event to all connected computers.

This shall not be used for beacons because the lore says you've no coordinates of the sender, and the server CPU would be consumed for no good reasons.

**Comment by nosknut** [comment]
Like wt i hear. Alltho. I think there should be receiver and transmitter block where receiver constantly drain a small amount of power since its only sniffing the messages dumped in subspace. As for distance messurement, it would be cool if people could make their own gps systrm. Aka transmitting receiving and comparing the time it took the message to get there. Delay logic like that would be very awsome

**Comment by nosknut** [comment]
Such a system would require it to be a universal time system. Possibly just pairing blocks for synchronizing their clocks.
That kind of gps system would not require to be run more then once per jump either btw ...

**Comment by LemADEC** [comment]
Having to craft the clocks together means you already have access to both blocks coordinates, which limit the interest for GPS. I would drop that constrains in favor of a subspace NTP.
GPS requires time precision with much greater resolution than the game tick rate, it can only be faked in the response.
Subspace lore has no positioning, so that GPS and time notion would only exist in the beacon system.

**Comment by nosknut** [comment]
U missunderstand. The gps is for ships that could get lost. Alltho comming to think of it, one could just grab the core position. My point was that if there is a delay per block so that a signal takes longer to travel large distances, that time could be used to calculate distance. Crafting clocks to be attached to a computer is a genius idea on ur side as it would allow you to implement some sort of atomic clock feature. I want to point out tho that wt i am suggesting is not a universal time, but a way of syncronizing the clocks of something. It would be a great addition to the mod since only clocks that have been synked would allow someone to find the position, allowing for some interresting creations and applications.if signal travels 2 blocks per tick, that would let the signal reach out 100k blocks in ruffly 40 mins. Just a suggestion but u get the idea...

**Comment by JohnNemECis** [comment]
I feel connected to this discussion now, so I'm going to drop some of my ideas here to. The range would be a good thing, but maybe it would be needed to make the antenna range big, but like in star trek (VOY), that the message could get scrambled when a little out of range, and completely invisible when really to far. maybe a +- 10% additional range in where there is interference, increasing a block further till almost none of the message will be visible, and at +-20% out of range, nothing to be found.

**Comment by nosknut** [comment]
I want to poibt out that irl a message would be traveling infinetly far for an infinite amount of time without getting weaker (in "empty" space). It is however redshifted over time due to the doplereffect. If u can compensate for the redshift and allow the signal to be received

**Comment by LemADEC** [comment]
No waves can travel indefinitely, for several reasons:
- space isn't empty, there's actual particles in there
- celestial objects might be on the way, most notably close to start and ending point
- even lasers aren't perfectly aligned and will slightly enlarge over a distance
- the distances between planets in the same start systems are huge, light takes approx. 6 mn from the sun to earth, way more to Pluto or the outer regions
- the distance between 2 star systems is several order of magnitude further 

In short, current technology can't do such transmissions.
Subspace lore is that's all broadcasted everywhere, independently of distance, but by nature, limited to small messages, since everyone is sharing that 'noise soup'. There's also a delay to decode the signal.
Radio based signals are much shorter distance but more rich and have some form of delay (beacon limited to same star system).

**Comment by nosknut** [comment]
#CMB

**Comment by LemADEC** [comment]
Cosmic microwave background?

**Comment by nosknut** [comment]
Yes
